### DISCUSSION TITLE
Suggestion: Templated Label read-only editor

### DISCUSSION THREAD

**Discussion by bjarnef** [leekelleher/umbraco-contentment#100] [Idea]
### Is your feature request related to a problem? Please describe.

Sometimes you want to store reference to external data in a property editor, e.g. using Label in core and properties "External Id" and "External Name". You could also store the raw JSON in Label (string), but doesn't look nice.

### Describe the solution you'd like

It could be useful to store an entity/object as JSON in a readonly/label editor and maybe a custom view/template to present to data, e.g.

```
Id: 1234
Name Unicorn
Category: Creatures
```

**Comment by leekelleher** [comment]
As a property-editor? 

Contentment does have a couple of read-only "editors" - ["readonly-node-preview.html"](https://github.com/leekelleher/umbraco-contentment/tree/develop/src/Umbraco.Community.Contentment/DataEditors/ReadOnly) is used on Content Blocks to show the selected ElementType.  [Notes](https://github.com/leekelleher/umbraco-contentment/tree/develop/src/Umbraco.Community.Contentment/DataEditors/Notes) is potentially reusable? or maybe [Render Macro](https://github.com/leekelleher/umbraco-contentment/tree/develop/src/Umbraco.Community.Contentment/DataEditors/RenderMacro)?

I did prototype a read-only editor using `<umb-code-snippet>` directive, but was one of those where I had no _real world_ use for it. 🤷

**Comment by bjarnef** [comment]
Yes, exactly 👌
I could be useful with some simple HTML formatting, e.g.

```
Name: <strong>{{model.name}}</strong><br>
Id: {{model.id}}
```

**Comment by leekelleher** [comment]
Okay, so a "Templated Label"?

Actually, [Notes](https://github.com/leekelleher/umbraco-contentment/tree/develop/src/Umbraco.Community.Contentment/DataEditors/Notes) isn't too far away from this, as the view can render AngularJS templates, (via my own `lk-bind-html-template` directive). The only limitation is that the DataType field uses TinyMCE, say instead of ACE. Hmmm, getting me thinking. 🤔

**Comment by leekelleher** [comment]
@bjarnef I've did a quick prototype.

![Screenshot 2021-04-20 174955](https://user-images.githubusercontent.com/209066/115434650-d7411b00-a200-11eb-84db-6f09ee4bbd23.png)

Swapped out one of my properties that had some Content Blocks data in there.

![Screenshot 2021-04-20 175037](https://user-images.githubusercontent.com/209066/115434797-02c40580-a201-11eb-95ff-ad7707acc6fc.png)

Turns out that I was able to reuse my Notes editor-view.  I only had to use 3 classes... and one of those is the ValueConverter, which is only there for playing nice with ModelsBuilder support.  I've put them in a Gist, if you wanted to play with it?

<https://gist.github.com/leekelleher/cfa4f3585ea0616241f22b3508f06e99>

I'm not 100% sold on the idea of included this as default with Contentment yet. It's one of those where I'd want to play with it more, incubate the idea, etc. _(I'm funny that way)_ 😆

**Reply by hfloyd** [reply]
@leekelleher I forked the repo, and added in those GISTs you provided... rebuilt, and now I have the "Templated Label" option 😎

One question, though... Is there some way I can get rid of the quotation marks so it does display nicely?

I used this template: 
`<umb-code-snippet language="'JSON'">{{model.value | json}}</umb-code-snippet>`

![image](https://user-images.githubusercontent.com/1252593/121433559-2e828280-c94a-11eb-8f35-d245844a1513.png)

Which does show the data... but not formatted (because of the value being a string):
![image](https://user-images.githubusercontent.com/1252593/121433644-49ed8d80-c94a-11eb-9078-5c2abe44669e.png)

**Reply by hfloyd** [reply]
I realized my mistake and changed it to:

`<umb-code-snippet language="'JSON'">{{model.value}}</umb-code-snippet>`

So now the extra quotation marks are gone... but still doesn't look nice like your example:
![image](https://user-images.githubusercontent.com/1252593/121434036-ba94aa00-c94a-11eb-9a30-6d5a03825371.png)

**Reply by bjarnef** [reply]
@hfloyd how about this:

`<umb-code-snippet language="'JSON'">{{control | json}}</umb-code-snippet>`

or 

`<umb-code-snippet language="'JSON'">{{block.data | json : 4 }}</umb-code-snippet>`

The component is used a few places in Umbraco core: https://github.com/umbraco/Umbraco-CMS/search?q=umb-code-snippet

**Reply by leekelleher** [reply]
It'll be because the values is stored as a serialized string of the JSON, so for the "pretty" (indented) formatting, it'd need to be deserialized. Something like `{{ JSON.parse(model.value) | JSON }}`. 

But I'm writing that from my mobile, excuse typos - or if it doesn't work. 😆

**Reply by hfloyd** [reply]
Okay, I'll try those, thanks 😊

Seems I have a more serious problem though...
When I have that property on the Video MediaType, then I select a new video, thus triggering the On_Saving event which actually grabs the JSON and puts it into the property - I'm getting an error:

```
System.InvalidOperationException: Cannot assign value "{"OEmbedStatus":2,"SupportsDimensions":true,"Markup":"<iframe width=\"1600\" height=\"900\" src=\"https://www.youtube.com/embed/3bSHnMZF9xI?feature=oembed\" frameborder=\"0\" allow=\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen></iframe>","FullJson":"{\"title\":\"Umbraco The Friendly CMS\",\"author_name\":\"Umbraco HQ\",\"author_url\":\"https://www.youtube.com/c/UmbracoHQ\",\"type\":\"video\",\"height\":900,\"width\":1600,\"version\":\"1.0\",\"provider_name\":\"YouTube\",\"provider_url\":\"https://www.youtube.com/\",\"thumbnail_height\":360,\"thumbnail_width\":480,\"thumbnail_url\":\"https://i.ytimg.com/vi/3bSHnMZF9xI/hqdefault.jpg\",\"html\":\"\\u003ciframe width=\\u00221600\\u0022 height=\\u0022900\\u0022 src=\\u0022https://www.youtube.com/embed/3bSHnMZF9xI?feature=oembed\\u0022 frameborder=\\u00220\\u0022 allow=\\u0022accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture\\u0022 allowfullscreen\\u003e\\u003c/iframe\\u003e\"}","OEmbedModel":{"Type":"video","Version":"1.0","Title":"Umbraco The Friendly CMS","author_name":"Umbraco HQ","author_url":"https://www.youtube.com/c/UmbracoHQ","provider_name":"YouTube","provider_url":"https://www.youtube.com/","thumbnail_url":"https://i.ytimg.com/vi/3bSHnMZF9xI/hqdefault.jpg","thumbnail_height":360.0,"thumbnail_width":480.0,"Html":"<iframe width=\"1600\" height=\"900\" src=\"https://www.youtube.com/embed/3bSHnMZF9xI?feature=oembed\" frameborder=\"0\" allow=\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen></iframe>","Url":null,"Height":900.0,"Width":1600.0},"IframeSrcUrl":"https://www.youtube.com/embed/3bSHnMZF9xI?feature=oembed","MarkupNoDimensions":"<iframe src=\"https://www.youtube.com/embed/3bSHnMZF9xI?feature=oembed\" frameborder=\"0\" allow=\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen=\"\"></iframe>","AspectRatio":{"ArCode":4,"Name":"16:9","Description":null,"Width":16,"Height":9,"Percentage":0.5625}}" 
of type "System.String" to property "OEmbedData" expecting type "System.Int32".
   at Umbraco.Core.Models.PropertyType.ThrowTypeException(Object value, Type expected, String alias)
   at Umbraco.Core.Models.PropertyType.TryConvertAssignedValue(Object value, Boolean throwOnError, Object& converted)
   at Umbraco.Core.Models.Property.SetValue(Object value, String culture, String segment)
   at Umbraco.Core.Models.ContentBase.SetValue(String propertyTypeAlias, Object value, String culture, String segment)
...
```

So, for some reason it thinks the property value data is an Int32? 😵

**Reply by leekelleher** [reply]
@hfloyd Ah yes, it's these two line...

- https://gist.github.com/leekelleher/cfa4f3585ea0616241f22b3508f06e99#file-templatedlabeldataeditor-cs-L41
- https://gist.github.com/leekelleher/cfa4f3585ea0616241f22b3508f06e99#file-templatedlabeldataeditor-cs-L59

When I tried out my prototype, I didn't have any extra data in the value, so it was an empty value for my testing.
If you change it to be `ValueTypes.Json` instead, it'll work for you. 

But in terms of a "proper" property-editor, the configuration would need the same as Umbraco's Label editor, with the dropdown to specify with underlying (SQL) data-type it is.

**Reply by hfloyd** [reply]
Thanks for that, @leekelleher 😁

Since this editor is focused on formatting, I don't know if other db types would be relevant.... Seems it would be long string.

**Reply by hfloyd** [reply]
@bjarnef This worked great:
`<umb-code-snippet language="'JSON'">{{model.value | json : 4 }}</umb-code-snippet>`

Look at this :
![image](https://user-images.githubusercontent.com/1252593/121535008-c0809e80-c9cf-11eb-852c-fdbe3d73c418.png)
❤❤❤👏

**Reply by leekelleher** [reply]
@hfloyd Excellent. 👏  Did you need to modify the `model.value`? or change the `ValueType`? or was this after a re-save? _Just curious._

**Reply by bjarnef** [reply]
@hfloyd great to know. I originally added the `<umb-code-snippet>` as we had a few code snippets in core https://github.com/umbraco/Umbraco-CMS/pull/7294 but also easier for packages to include code snippets in a consistent way. Now the Block List editor also use this in unsupported block view. It could be enhanced with some code highlighting. At that time I considered [PrismJS](https://prismjs.com/) but we already have syntax highlighting with the Ace editor.

@warrenbuckley also did this lab project using Monaco editor: https://github.com/umbraco/Umbraco-CMS/pull/7524

Could be something to consider with the new BackOffice UI later 👍 

I think the one from the Microsoft documentation is pretty nice: https://docs.microsoft.com/en-us/odata/webapi/netcore

**Reply by hfloyd** [reply]
@leekelleher : The changes I made were the ones you suggested :
![image](https://user-images.githubusercontent.com/1252593/121539086-58cc5280-c9d3-11eb-81e6-2f691cbabb4e.png)

Here is a little demo of the total awesomeness!

![Oembed-Video](https://user-images.githubusercontent.com/1252593/121539491-ba8cbc80-c9d3-11eb-99d6-c22c5d656e29.gif)


Massive #H5YR to you both! 💖

**Reply by leekelleher** [reply]
@hfloyd Excellent! 🙌   One thing that I've been doing if the JSON/code snippets are getting too long, is to wrap the `<umb-code-snippet>` in a `<details><summary>` tags. Give's that little toggle, if you need to condense the vertical space.

**Reply by hfloyd** [reply]
ooo... let me try that!

**Reply by leekelleher** [reply]
@hfloyd You can apply Umbraco backoffice styles it too, e.g.

```html
<details class="well well-small">
  <summary><strong>View the raw oEmbed JSON data</strong></summary>
  <umb-code-snippet language="'JSON'">{{model.value | json : 4 }}</umb-code-snippet>
</details>
```

Other classes could be `alert alert-info`, or the other alerts, `alert-success`, `alert-warning`, `alert-error`, I think there might even be an `alert-form` one. Anyway, worth playing around with.

**Reply by hfloyd** [reply]
Sexy!
![image](https://user-images.githubusercontent.com/1252593/121544002-5e2b9c00-c9d7-11eb-85f5-016460ab700f.png)

**Reply by hfloyd** [reply]
So.... @leekelleher Does this now rank for inclusion? 😉

**Comment by hfloyd** [comment]
You know, I think I'd like something similar to this as well. The main use-case is that I frequently have a need to pull in some external data from an API call or whatever and store it with the node data. Sometimes this is done with an event ("OnSave"), and sometimes the data is being updated by an external operation - like a daily import or something.

What I have found is that if I use a "Label" with the "Long string" value type, I can't access the stored data (it always comes across as "NULL" in models, etc.), so I tend to use the "Textarea" - which works.

The challenge is that it is visible - and editable - to the Editors, and I don't want them to edit the Serialized JSON. (With the possible exception of a "Clear" button, so the OnSave event could regenerate it.)

I don't personally need a fancy templated display. I think your formatted JSON example is plenty readable. 😊

My "dream" PropertyEditor would allow some config, perhaps like this:

![image](https://user-images.githubusercontent.com/1252593/121394628-8a83e180-c91f-11eb-829c-a444ca27daa2.png)


One thing I have been working on is a better way to manage external videos (YouTube, Vimeo, etc.), via a "Video" media type, which can then be "picked" on Content nodes, etc. I came across [Dave's "OEmbed-Picker-Property-Editor"](https://github.com/dawoe/OEmbed-Picker-Property-Editor/), which makes the RTE "Embed" tool available as a stand-alone property. Looking into the OEmbed data which is available from these services, I realized that I wanted to have more than just the embed HTML available to me, so I wrote an event which duplicates some of the "Embed" tool functionality to do it's own call to the oembed api endpoint and pull back the full set of available data, which then gets processed a bit into a custom object model, and stored in as serialized JSON on the Video media node - right now in a Textarea:

![image](https://user-images.githubusercontent.com/1252593/121426216-55888680-c941-11eb-89c1-97026b019f2d.png)

It would be awesome if I could "protect" this data - and even display it to the Content Editor in a nicer format, since they can see it anyway...

**Reply by bjarnef** [reply]
@hfloyd you may want to have a look at the editor model to hide the property to editors - or only show this to admin users: https://our.umbraco.com/documentation/reference/events/EditorModel-Events/

I use something like this on a project. In this case it doesn't use variants, but in case the project does use variants, you would need to loop these.

```
using ContentModels = Umbraco.Web.PublishedModels;
...

[RuntimeLevel(MinLevel = RuntimeLevel.Run)]
public class SubscribeToEditorModelEventsComposer : ComponentComposer<SubscribeToEditorModelEvents>
{
    //this automatically adds the component to the Components collection of the Umbraco composition
}

public class SubscribeToEditorModelEvents : IComponent
{
    public SubscribeToEditorModelEvents() { }

    // Initialize: runs once when Umbraco starts
    public void Initialize()
    {
        EditorModelEventManager.SendingContentModel += EditorModelEventManager_SendingContentModel;
    }

    // Terminate: runs once when Umbraco stops
    public void Terminate()
    {
        // Unsubscribe during shutdown
        EditorModelEventManager.SendingContentModel -= EditorModelEventManager_SendingContentModel;
    }

    private void EditorModelEventManager_SendingContentModel(HttpActionExecutedContext sender, EditorModelEventArgs<Umbraco.Web.Models.ContentEditing.ContentItemDisplay> e)
    {
        var userGroupAliasses = e.UmbracoContext.Security.CurrentUser.Groups.Select(userGroup => userGroup.Alias).ToList();

        if (e.Model.ContentTypeAlias == ContentModels.Settings.ModelTypeAlias)
        {
            if (userGroupAliasses.Any(alias => alias.Equals(Constants.Security.AdminGroupAlias)) == false)
            {
                // Hide specific properties if not admin.
                HideProperties(e.Model, userGroupAliasses,
                    new List<string>
                    {
                        { ContentModels.Settings.GetModelPropertyType(x => x.MyFirstProperty).Alias },
                        { ContentModels.Settings.GetModelPropertyType(x => x.MySecondProperty).Alias }
                    }
                );

                // Hide tabs with no further properties.
                HideEmptyTabs(e.Model);
            }
        }
    }

    private static void HideProperties(ContentItemDisplay contentItemDisplay, List<string> userGroupAliasses, List<string> propertyAliases)
    {
        if (propertyAliases.Any())
        {
            foreach (var tab in contentItemDisplay.Variants.First().Tabs)
            {
                tab.Properties = tab.Properties.Where(property => !propertyAliases.Contains(property.Alias));
            }
        }
    }

    private static void HideEmptyTabs(ContentItemDisplay contentItemDisplay)
    {
        contentItemDisplay.Variants.First().Tabs = contentItemDisplay.Variants.First().Tabs.Where(tab => tab.Properties.Any());
    }
}
```

Regarding the JSON format, maybe the [Code Editor](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/code-editor.md) can be used?

**Reply by hfloyd** [reply]
Thanks for the tip, @bjarnef ! If I can't get something like this "templated label" working the way I want, I will look into this as well.

**Comment by leekelleher** [comment]
Good news... I've opened PR #133! 🎉 

I need to complete the documentation before merging it in. (Or if anyone wants to help, it's open to any help) 😄 

Interesting part in this Template Label editor is the "Developer mode" toggle (which I built for some of the other Contentment editors), meaning that you can make edits to the underlying data once it is in there.

**Comment by leekelleher** [comment]
...and it's been released in [Contentment v2.2.0](https://github.com/leekelleher/umbraco-contentment/releases/tag/2.2.0)! 🌟

**Reply by bjarnef** [reply]
@leekelleher will it also be included in v3 as well? 😁 🚀

**Reply by leekelleher** [reply]
💯 100% yes, just need to find time to do all the merging and releasing, (it's taken me 3 days to find time to do the 2.2.0 release, eek! 🙀)

With v3, there are changes coming soon... as I've decided to take the .NET multi-targeting approach - I found that supporting separate branches would become a maintenance headache. 😅

**Reply by leekelleher** [reply]
...and it's in [Contentment v3.0.0-alpha006](https://github.com/leekelleher/umbraco-contentment/releases/tag/3.0.0-alpha006)!
This isn't the multi-targeted one, that'll take me a while longer. 😆

**Comment by leekelleher** [comment]
Closing this off, as the idea became the [Templated Label editor](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/templated-label.md).
